core/src/sumeragi/main_loop.rs

```diff
@@ -448,60 +413,71 @@ fn replace_top_block<F: FaultInjection>(
     block: impl Into<VersionedCommittedBlock>,
 ) {
     let committed_block = block.into();
-    let block_hash = committed_block.hash();
 
     state.wsv = state.finalized_wsv.clone();
-    state
-        .wsv
-        .apply(&committed_block)
-        .expect("Failed to apply block on WSV. Bailing.");
-
-    sumeragi.send_events(state.wsv.events_buffer.replace(Vec::new()));
-
-    // Update WSV copy that is public facing
-    *sumeragi.wsv.lock() = state.wsv.clone();
-
-    // This sends "Block committed" event, so it should be done
-    // AFTER public facing WSV update
-    sumeragi.send_events(&committed_block);
-
+    update_state(state, sumeragi, &committed_block);
     // state.previous_block_hash stays the same.
-    state.latest_block_height = committed_block.header().height;
-    state.latest_block_hash = Some(block_hash);
-    state.latest_block_view_change_index = committed_block.header().view_change_index;
-
-    let current_topology = &mut state.current_topology;
-    let role = current_topology.role(&sumeragi.peer_id);
 
     info!(
-        addr=%sumeragi.peer_id.address, %role,
+        addr=%sumeragi.peer_id.address,
+        role=%state.current_topology.role(&sumeragi.peer_id),
         block_height=%state.latest_block_height,
-        %block_hash, "Replacing top block"
+        block_hash=%committed_block.hash(),
+        "Replacing top block"
     );
 
-    *current_topology = Topology {
+    update_topology(state, &committed_block);
+
+    sumeragi.kura.replace_top_block(committed_block);
+
+    cache_transaction(state, sumeragi)
+}
+
+fn update_topology(state: &mut State, committed_block: &VersionedCommittedBlock) {
+    let mut topology = Topology {
         sorted_peers: committed_block.header().committed_with_topology.clone(),
     };
-    current_topology.lift_up_peers(
+    topology.lift_up_peers(
         &committed_block
             .signatures()
             .into_iter()
             .map(|s| s.public_key().clone())
             .collect::<Vec<PublicKey>>(),
     );
-    current_topology.rotate_set_a();
-    current_topology.update_peer_list(
+    topology.rotate_set_a();
+    topology.update_peer_list(
         &state
             .wsv
             .peers_ids()
             .iter()
             .map(|id| id.clone())
             .collect::<Vec<PeerId>>(),
     );
+    state.current_topology = topology;
+}
 
-    sumeragi.kura.replace_top_block(committed_block);
+fn update_state<F: FaultInjection>(
```

Comment:
This refactor doesn't seem to do much other than extract the logic into a single function. The function isn't particularly reusable either. I'm not against merging it, but I'd think that a dedicated refactor PR should do a little more, what do you think?

Reply:
@appetrosyan what direction should be taken in your opinion?
Intention was to extract reused functionality to separate functions to avoid duplication and divergence in behavior.

I tried to use `const generic bool` value but than code is polluted with `if` statements which doesn't look very nice to me.

Reply:
> I tried to use const generic bool value but than code is polluted with if statements which doesn't look very nice to me.

Those aren't particularly nice looking in rust. In C++ it's a different situation. 

> Intention was to extract reused functionality (state update and topology update) to separate functions to avoid duplication and divergence in behavior.

That's a good endeavour, and we should merge it, my question is whether we could expand the scope of this refactoring to more aspects of this same file? Compress it somewhat?

Reply:
Yeah, i absolutely agree that some tests for sumeragi is would be beneficial. Do you mind separate ticket for this?

Reply:
#3302